### ISSUE TITLE
Restart build immediately if a file changes while the app is building

### ISSUE THREAD

**Issue by trusktr** [meteor/meteor-feature-requests#93]
### Feature requests should include as much detail as possible:
- [x] A descriptive title
- [x] A description of the problem you're trying to solve, including *why* you think this is a problem
- [ ] An overview of the suggested solution
- [ ] If the feature changes current behavior, reasons why your solution is better

Original description:

When a client file changes, a rebuild happens, and eventually the client will be restarted.

But, if we change another client file while a build is in progress (the client is yet to be restarted) then that will trigger another build that queues up another client restart.

Idea: when a file changes during a build, cancel the following restart and only restart a single time once files have stopped changing; similar to how _.debounce works.

I find myself editing multiple files and saving them in succession, which causes multiple restarts. A single restart would be more efficient. The timeout of the "debounce" can just be the amount of time it takes for a rebuild to complete.

**Comment by trusktr** [comment]
@stubailo said:

This is a great idea! Happens to me all the time.

Thanks for the feature request! We welcome discussions about how to make Meteor better. If you haven't yet, check out our wiki page about feature requests on GitHub.

**Comment by trusktr** [comment]
I don't know what I was thinking about the denounce idea though. The current build should just be cancelled, and a new one started when files change during a build. It should happen as quickly as possible, and be cancelable mid-build without causing any problems.

**Comment by stubailo** [comment]
Yeah, there's probably not much value to finishing the initial build since you'll just be looking at old code!

**Comment by benjamn** [comment]
Since Meteor 1.4.2, thanks to aggressive caching within the build system, it's less obvious this idea would meaningfully improve the developer experience, since it may be better to let the current build finish, so that as much work as possible is cached, then check once for changed files using `watch.isUpToDate` at the end, and kick off a quick incremental rebuild if anything changed.

Put another way, if you have to do two rebuilds, it doesn't really help to move work from the first rebuild to the second, as long as work is not being duplicated between the builds.

There's definitely value in not restarting the server until the rebuilding has settled down, but we already do that. Another idea would be to skip minification if there's been a file change, but minification typically does not happen during development.

I don't mean to shut this idea down. I get the intuitive appeal of something like this. It made a lot of sense to me, too, pre-Meteor 1.4.2. Meteor works a bit differently now, so I hope your intuitions will incorporate those details.

**Comment by trusktr** [comment]
Cool, sounds like it is better. The main problem I had before was that if I saved three separate files, it made three new builds happen, and they all took the same. Maybe it is better now though, I'll keep an eye out. I simply copies this issue over, which may be outdated. I can reopen if I notice anything like the old problem.

**Comment by benjamn** [comment]
What I said above is more true than before Meteor 1.4.2, though it may not be totally true in all cases yet. However, I would rather fix the caching logic to make incremental rebuilds as cheap as possible, instead of attempting to cancel rebuilds.

**Comment by trusktr** [comment]
Sounds good, and also making sure that three file changes (for example) results in one incremental build rather than three new ones (not sure if this may already be fixed, will pay attention now that I'm back in Meteor app dev for a while).

**Comment by trusktr** [comment]
Alright, so I've observed for a while on Meteor `1.5-rc.0`, and what I've noticed is that if I make at least two file changes, then the client will refresh twice, which is somewhat awkward for the dev experience.

What happens is the first file change starts a rebuild and will cause a refresh. Then if one or more files were modified during that first rebuild, a second rebuild will happen after, which causes the client to refresh again.

There will only be a third refresh if one or more files are saved during the second rebuild.

It'd be nice for Meteor to detect a file change during rebuild, and even if it will do another build, at least just skip the client refresh to wait for the next build to finish.

**Comment by benjamn** [comment]
There's definitely a set of things that should not happen until the rebuilding has settled down, and refreshing the client seems like one of them.
